src/test/java/dev/blaauwendraad/masker/json/MaskingStateTest.java

```diff
@@ -91,4 +92,82 @@ void shouldUseCorrectOffsetWhenThrowingValueMaskerError() {
+    void shouldReadInitialBuffer() {
+        String json = "[" + "{\"maskMe\":\"val\"},".repeat(2048) + "0]";
+        MaskingState maskingState = new MaskingState(new ByteArrayInputStream(json.getBytes(StandardCharsets.UTF_8)),
+                new ByteArrayOutputStream(), false);
+
+        Assertions.assertThat(maskingState.messageLength).isEqualTo(8192);
+        Assertions.assertThat(maskingState.getMessage()).hasSize(8192);
+    }
+
+    @Test
+    void shouldReadNewBufferWithoutCurrentValueStartIndex() {
+        String json = "[" + "{\"maskMe\":\"val\"},".repeat(2048) + "0]";
+        MaskingState maskingState = new MaskingState(new ByteArrayInputStream(json.getBytes(StandardCharsets.UTF_8)),
+                new ByteArrayOutputStream(), false);
+
+        for (int i = 0; i < 8192; i++) {
+            maskingState.next();
+        }
+
+        Assertions.assertThat(maskingState.messageLength).isEqualTo(8192);
+        Assertions.assertThat(maskingState.getMessage()).hasSize(8192);
+    }
+
+    @Test
+    void shouldMoveCurrentValueAndReadNewBuffer() {
+        String json = "[" + "{\"maskMe\":\"val\"},".repeat(2048) + "0]";
+        MaskingState maskingState = new MaskingState(new ByteArrayInputStream(json.getBytes(StandardCharsets.UTF_8)),
+                new ByteArrayOutputStream(), false);
+
+        for (int i = 0; i < 8181; i++) {
+            maskingState.next();
+        }
+        byte[] currentValue = new byte[10];
+        maskingState.registerValueStartIndex();
+        currentValue[0] = maskingState.byteAtCurrentIndex();
+        for (int i = 0; i < 10; i++) {
+            currentValue[i] = maskingState.byteAtCurrentIndex();
+            maskingState.next();
+        }
+
+        maskingState.readNextBuffer();
+
+        for (int i = 0; i < 10; i++) {
+            Assertions.assertThat(maskingState.getByte(i)).isEqualTo(currentValue[i]);
+        }
+        Assertions.assertThat(maskingState.getMessage()).hasSize(8192);
+    }
+
+    @Test
+    void shouldExtendCurrentBuffer() {
+        String json = "[" + "{\"maskMe\":\"val\"},".repeat(2048) + "0]";
+
+        MaskingState maskingState = new MaskingState(new ByteArrayInputStream(json.getBytes(StandardCharsets.UTF_8)),
+                new ByteArrayOutputStream(), false);
+
+        for (int i = 0; i < 4096; i++) {
+            maskingState.next();
+        }
+        maskingState.registerValueStartIndex();
+        for (int i = 0; i < 4095; i++) {
+            maskingState.next();
+        }
+        maskingState.readNextBuffer();
+
+        Assertions.assertThat(maskingState.getMessage()).hasSize(10240);
+    }
+
+    @Test
+    void shouldReachEndOfStream() {
+        String json = "{\"maskMe:\":\"val\"}";
+
+        MaskingState maskingState = new MaskingState(new ByteArrayInputStream(json.getBytes(StandardCharsets.UTF_8)),
+                new ByteArrayOutputStream(), false);
+
+        Assertions.assertThat(maskingState.messageLength).isEqualTo(17);
+    }
```

Comment:
Maybe I'm missing it, but are there tests to cover these scenarios?
1. Extending buffer when the (matching) key is in between the first quarter and the rest of the buffer
2. Extending buffer when the (matching) key is fully in the last 3/4 of the buffer
3. Extending buffer when the (matching) key is larger than 2 buffers sizes
4. Extending buffer when the (masked) value is in between the first quarter and the rest of the buffer
5. Extending buffer when the (masked) value is fully in the last 3/4 of the buffer
6. Extending buffer when the (masked) value is larger than 2 buffer sizes
7. Using `ValueMaskers.withTextValue` on value that extends the buffer (to verify that all `ValueMaskerContext` methods work correctly)

Reply:
This is what we had in mind as well, and we think it would be easier if you can configure the buffer size to cover all these cases.

Reply:
Agree, I will make the buffer size package-private and configure it small enough in tests so that I can cover these scenarious.

Reply:
Done, created a separate test suit for that.